File: app/src/main/java/com/buzbuz/dikkeneksoundboard/SoundDatabaseHelper.java
package com.buzbuz.dikkeneksoundboard;

import android.content.ContentValues;
import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.provider.BaseColumns;
import android.util.Log;

import java.io.IOException;

/**
 * Class offering helping functions to access and modify the sound database
 */
public class SoundDatabaseHelper extends SQLiteOpenHelper {

    /** Tag for logs */
    private static final String LOG_TAG = "SoundDatabaseHelper";

    /** Helper static class containing all columns available in the database */
    public static abstract class SoundColumn implements BaseColumns {
        /** Column containing the name of the sound used for display purpose */
        public static final String COLUMN_NAME_TITLE = "_display_name";
        /** Column containing the path of the file in the assets of the app */
        public static final String COLUMN_NAME_PATH = "path";
        /** Column indicating if the sound is in the user favourites or not */
        public static final String COLUMN_NAME_FAVOURITE = "favourites";
        /** Nullable column (see Android doc) */
        public static final String COLUMN_NAME_NULLABLE = "nullable";

        /** Array of all columns available in the database */
        public static final String[] COLUMNS_LIST = {
                _ID,
                COLUMN_NAME_TITLE,
                COLUMN_NAME_PATH,
                COLUMN_NAME_FAVOURITE,
                COLUMN_NAME_NULLABLE
        };
    }
    /** Name of the table containing all sounds */
    public static final String TABLE_NAME = "sounds";

    private static final String TEXT_TYPE = "TEXT";
    private static final String INTEGER_TYPE = "INTEGER";
    private static final String SQL_CREATE_DATABASE =
            "CREATE TABLE " + TABLE_NAME + " (" +
                    SoundColumn._ID + " INTEGER PRIMARY KEY," +
                    SoundColumn.COLUMN_NAME_TITLE + " " + TEXT_TYPE + "," +
                    SoundColumn.COLUMN_NAME_PATH + " " + TEXT_TYPE + "," +
                    SoundColumn.COLUMN_NAME_FAVOURITE + " " + INTEGER_TYPE +
            " )";

    private static final String DATABASE_NAME = "sounds.db";
    private static final int DATABASE_VERSION = 1;

    private Context mContext;

    public SoundDatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
        mContext = context;
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.execSQL(SQL_CREATE_DATABASE);
        populate(db);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // Not needed as they will be no newer sounds added to database
    }

    /**
     * List all sounds files from the assets and fill the database with it.
     * Must be called at first launch of app
     */
    private void populate(SQLiteDatabase db) {
        try {
            String[] assetsList = mContext.getAssets().list("");
            if (assetsList.length > 0) {
                for (String soundFile : assetsList) {
                    if (soundFile.endsWith(".mp3")) {
                        ContentValues rowValues = new ContentValues();
                        rowValues.put(SoundColumn.COLUMN_NAME_TITLE, formatSoundTitle(soundFile));
                        rowValues.put(SoundColumn.COLUMN_NAME_PATH, soundFile);
                        rowValues.put(SoundColumn.COLUMN_NAME_FAVOURITE, 0);
                        db.insert(TABLE_NAME, SoundColumn.COLUMN_NAME_NULLABLE, rowValues);
                    }
                }
            } else {
                Log.e(LOG_TAG, "Can't find any sound files from assets root folder");
            }
        } catch (IOException ex) {
            Log.e(LOG_TAG, "Can't get sound list from assets");
        }
    }

    private String formatSoundTitle(String filePath) {
        String result;
        result = filePath.substring(0, filePath.length()-4);
        result = result.replace('_', ' ');
        return result;
    }
}

File: app/src/main/java/com/buzbuz/dikkeneksoundboard/provider/SoundDatabaseHelper.kt
/*
 * Copyright (C) 2020 Nain57
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; If not, see <http://www.gnu.org/licenses/>.
 */
package com.buzbuz.dikkeneksoundboard.provider

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.database.sqlite.SQLiteQueryBuilder
import android.util.Log

/** SQLiteOpenHelper implementation managing the SQLite database containing all sounds informations. */
class SoundDatabaseHelper(private val context: Context) :
    SQLiteOpenHelper(context, "sounds.db", null, 1) {

    companion object {
        /** Tag for logs */
        private const val TAG = "SoundDatabaseHelper"

        /** The name of the table in the database containing all sounds. */
        private const val TABLE_NAME ="sounds"
        /** SQL request used to create the table of sounds in the database. */
        private const val SQL_CREATE_DATABASE = "CREATE TABLE $TABLE_NAME (" +
                "${SoundColumns.COLUMN_ID} INTEGER PRIMARY KEY," +
                "${SoundColumns.COLUMN_NAME_TITLE} TEXT," +
                "${SoundColumns.COLUMN_NAME_PATH} TEXT," +
                "${SoundColumns.COLUMN_NAME_FAVOURITE} INTEGER )"
    }

    override fun onCreate(db: SQLiteDatabase?) {
        Log.i(TAG, "Creating database")

        db!!.execSQL(SQL_CREATE_DATABASE)

        val soundPaths = context.assets.list("")?.filter { it.endsWith(".mp3") }
        if (!soundPaths.isNullOrEmpty()) {
            for (soundPath in soundPaths) {
                val value = ContentValues()
                value.put(SoundColumns.COLUMN_NAME_TITLE, formatSoundTitle(soundPath))
                value.put(SoundColumns.COLUMN_NAME_PATH, soundPath)
                value.put(SoundColumns.COLUMN_NAME_FAVOURITE, 0)
                db.insert(TABLE_NAME, null, value)
            }
        }
    }

    override fun onUpgrade(db: SQLiteDatabase?, oldVersion: Int, newVersion: Int) {
        // Not needed as they will be no newer sounds added to database
    }

    /**
     * Perform a query on the sound database.
     * <p>
     * @param projection a list of which columns to return.
     * @param selection filter declaring which rows to return, formatted as an SQL WHERE clause (excluding the WHERE
     *                  itself). Passing null will return all rows for the given URL.
     * @param selectionArgs you may include ?s in selection, which will be replaced by the values from selectionArgs,
     *                      in order that they appear in the selection. The values will be bound as Strings.
     * @param sortOrder how to order the rows, formatted as an SQL ORDER BY clause (excluding the ORDER BY itself).
     *                  Passing null will use the default sort order, which may be unordered.
     * <p>
     * @return a cursor over the result set
     */
    fun queryDatabase(
        projection: Array<out String>,
        selection: String?,
        selectionArgs: Array<out String>?,
        sortOrder: String?
    ): Cursor {

        val queryBuilder = SQLiteQueryBuilder()
        queryBuilder.tables = TABLE_NAME
        return queryBuilder.query(writableDatabase, projection, selection, selectionArgs, null, null, sortOrder)
    }

    /**
     * Format the title of a sound using its path.
     * <p>
     * @param path the path of the sound to format from.
     * <p>
     * @return the formatted title.
     */
    private fun formatSoundTitle(path: String) : String = path.substring(0, path.length - 4).replace('_', ' ')
}